probe-rs/src/architecture/riscv/mod.rs

```diff
@@ -0,0 +1,664 @@
+            dmcontrol.set_haltreq(true);
+
+            self.interface.write_dm_register(dmcontrol)?;
+
+            log::debug!("Clearing ndmreset bit");
+            let mut dmcontrol = Dmcontrol(0);
+            dmcontrol.set_dmactive(true);
+            dmcontrol.set_ndmreset(false);
+            dmcontrol.set_haltreq(true);
+
+            self.interface.write_dm_register(dmcontrol)?;
+        }
+
+        // check that cores have reset
+        let readback: Dmstatus = self.interface.read_dm_register()?;
+
+        if !(readback.allhavereset() && readback.allhalted()) {
+            return Err(RiscvError::RequestNotAcknowledged.into());
+        }
+
+        // acknowledge the reset, clear the halt request
+        let mut dmcontrol = Dmcontrol(0);
+        dmcontrol.set_dmactive(true);
+        dmcontrol.set_ackhavereset(true);
+
+        self.interface.write_dm_register(dmcontrol)?;
+
+        let pc = self.read_core_reg(CoreRegisterAddress(0x7b1))?;
+
+        Ok(CoreInformation { pc })
+    }
+
+    fn step(&self) -> Result<crate::core::CoreInformation, crate::Error> {
+        let mut dcsr = Dcsr(self.read_core_reg(CoreRegisterAddress(0x7b0))?);
+
+        dcsr.set_step(true);
+
+        self.write_csr(0x7b0, dcsr.0)?;
+
+        self.run()?;
+
+        self.wait_for_core_halted()?;
+
+        let pc = self.read_core_reg(CoreRegisterAddress(0x7b1))?;
+
+        // clear step request
+        let mut dcsr = Dcsr(self.read_core_reg(CoreRegisterAddress(0x7b0))?);
+
+        dcsr.set_step(false);
+
+        self.write_csr(0x7b0, dcsr.0)?;
+
+        Ok(CoreInformation { pc })
+    }
+
+    fn read_core_reg(&self, address: crate::CoreRegisterAddress) -> Result<u32, crate::Error> {
+        // We need to sue the "Access Register Command",
+        // which has cmdtype 0
+
+        // write needs to be clear
+        // transfer has to be set
+
+        log::debug!("Reading core register at address {:#x}", address.0);
+
+        // if it is a gpr (general purpose register) read using an abstract command,
+        // otherwise, use the program buffer
+        if address.0 >= 0x1000 && address.0 <= 0x101f {
+            let value = self
+                .interface
+                .abstract_cmd_register_read(address.0 as u32)?;
+            Ok(value)
+        } else {
+            let reg_value = self.read_csr(address.0)?;
+            Ok(reg_value)
+        }
+    }
+
+    fn write_core_reg(
+        &self,
+        address: crate::CoreRegisterAddress,
+        value: u32,
+    ) -> Result<(), crate::Error> {
+        if address.0 >= 0x1000 && address.0 <= 0x101f {
+            self.interface
+                .abstract_cmd_register_write(address.0 as u32, value)?;
+        } else {
+            self.write_csr(address.0, value)?;
+        }
+        Ok(())
+    }
+
+    fn get_available_breakpoint_units(&self) -> Result<u32, crate::Error> {
+        // TODO: This should probably only be done once, when initialising
```

Comment:
Yep. Also consider fetching more info like available progbuf slots, features etc... all somewhere at initialization.

Reply:
I'd prefer to add this later, so this PR can be merged. I've opened #152 to track this.